all_equal

davidlee1102 opened the pull request:
Hi,
Please check on my code and if you have any advices, feel free to talk with me 
Thank you,
David

davidlee1102 commented:
Hi @VedPatwardhan ,
Nice to see you again 😸 , hope that this code could be fine, if you have any advices, please let me know,
Thank you,
David

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, thanks for contributing and nice to see you again! There are few changes to be made to your PR before we get it merged:
1. Instead of `Any` input, please add 3 sections naming with 'int`, `float`, `tuple`, `list`, `ivy.Array` and `ivy.NativeArray` inputs as that makes it better for developers to see that any types are supported. 
2. For the mix input examples, you could do a comparison between an `ivy.Container` and an `ivy.Array` or an `ivy.Container` and an `ivy.NativeArray`, so please remove the `With a mix of :code:`ivy.Container` and :code:`Any` input:` and replace them with the above ones. You could refer [this](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html) page so that uniformity is maintained in the sequence of examples.

Feel free to ask any questions! Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
Thank you for giving me some advices, here is my progress about your advices:
1. I have divided into 3 categories (int, float) (tuple, list) and (ivy.Array, ivy.NativeArray) input
2. I have separated into 2 forms of this ivy.Container (ivy.Array, ivy.NativeArray) input
3. I have changed the name of this
So please check that, if you need to change something, please let me know
Thank you,
David

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, I'd suggest the following changes to your PR:
1. You don't need to add a grid search over all possible combinations of the parameters, so please restrict the `With a mix of int and float input:` examples to 2.
2. Please add one example each for `With int input`, `With float input`, `With ivy.Array input` and `With ivy.NativeArray input`. I'm not referring to a mix of the input types but for comparing two inputs of the **same type**.
3. For `With a mix of ivy.Array and ivy.NativeArray` and `With a mix of tuple and list` too, please restrict the number of examples to 2 each. Basically, you don't need to have examples showing two possible `equality_matrix` values for each input combination. You can find more about this [here](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html).

Let me know if you have any questions. Thanks 🙂

davidlee1102 commented:
Thank you, 
I have fixed all problems you have told me, so it's better if you can check my code again

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, one final change, `With a mix of ivy.Array and ivy.Container inputs:` means one input should be an `ivy.Container` and the other should be an `ivy.Array`. In your case it seems that you have both inputs as `ivy.Container`, please update that. Same applies to `With a mix of ivy.NativeArray and ivy.Container` too.
Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
I have fixed some errors, please check on my code
Another problem I saw that was many errors came from other file, it did not caused by my code 
Thank you 😺 
David

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, as per the new guidelines for writing the docstrings, the examples won't be all in one place, but will be kept in different files depending on whether it's a functional array example, a functional container example or an instance example, you can refer the updated [guide](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html) for the same. A message regarding it was also recently circulated on discord. Please update the docstrings accordingly.
Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
I understand what you said to me, however, I did not see where a functional container example, so could you instruct me where I put it on?
Thanks,
David

VedPatwardhan commented:
> Hi @VedPatwardhan , I understand what you said to me, however, I did not see where a functional container example, so could you instruct me where I put it on? Thanks, David

All examples called in the form `ivy.all_equal` should be in the same file as now i.e. the functional/ivy folder 🙂

davidlee1102 commented:
Hi @VedPatwardhan , 
Can you instruct me how to do with that? I have tried to read the docstring [https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html](url) but I dont understand where I can start for putting this code
Thanks
David

davidlee1102 commented:
Hi @VedPatwardhan ,
I have reverted the original code (All examples called in the form ivy.all_equal have already inside general.py ), could you check it again?
Thank you
David

VedPatwardhan reviewed (changes requested):
1. The examples using the `ivy.Array` instance method should be added to [this](https://github.com/unifyai/ivy/blob/36962661753ba56aee2487ef4e96c5e7b130f162/ivy/array/general.py) file in your case and the `ivy.Container` instance method should be added to [this](https://github.com/unifyai/ivy/blob/36962661753ba56aee2487ef4e96c5e7b130f162/ivy/container/general.py) file.
2. In case the function isn't implemented in the respective files, you have to implement it in those files and then add the docstring there.

Let me know if you have any questions, thanks 🙂

davidlee1102 commented:
@VedPatwardhan ,
Thank for suggesting me and supporting me for updating new code following the new contribution roles
Please check on that and give me feedback if you have
Thank you again,
David

VedPatwardhan reviewed (changes requested):
You're editing the correct files, however there are a few changes:
1. For `ivy.Array` or `ivy.Container` instance methods, we should be having the first argument as `self` which will be the `ivy.Array` or `ivy.Container` for which the method is called. The second argument can be of any type, but the first argument would be `self` which is an `ivy.Array` or `ivy.Container` depending on the method concerned.
2. We shouldn't be implementing the functionality inside the method, we should be making use of the functional API function we have already implemented. For e.g. the `ivy.Array` instance method should be returning `ivy.all_equal(self, x)` inside the instance method for `ivy.Array`. 
3. In case of `ivy.Container`, it's a bit different. We can discuss that after 2 is done.

Let me know if you have any questions. Thanks 🙂

davidlee1102 commented:
@VedPatwardhan ,
Thank you for supporting me to do this issue, I have fixed based your instruction, let check and move to 3. of your comment
David

VedPatwardhan commented on ivy/array/general.py:
```diff
@@ -1,8 +1,77 @@
+        True
+
+        """
+        return ivy.all_equal(self, equality_matrix)
```
The function should be called in the form `ivy.all_equal(self, x, equality_matrix)` where `x` is the array or list or anything that is compared to our array `arr` in the form `arr.all_equal(x, equality_matrix=<True|False>)` by the user.

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102:
1. Done, thanks for making the changes
2. The `ivy.Array` instance method needs to be updated, I have added more details near that line.
3. Please update the docstrings in the `ivy.Array` and `ivy.Container` instance method implementations. They should only be having examples called in the form `a.all_equal(b, equality_matrix=True)` where `a` could be an `ivy.Array` or `ivy.Container` respectively. 
4. The `ivy.Container` instance method also has a similar structure, but it has other details like adding a static method along with an instance method, etc. I think it's better if we look at it after the above changes are completed.

Let me know if you have any questions. Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
I have fixed all things you have told me about that, so let check again and see what I need to fix next
Thank you
David

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, Thanks for making the changes please also update the docstrings as mentioned in the previous review. Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
Let check again, thank you
David

VedPatwardhan commented on ivy/array/general.py:
```diff
@@ -1,8 +1,77 @@
 class ArrayWithGeneral(abc.ABC):
-    pass
+    def all_equal(
+            self,
```
please add `x` too as a parameter which is the parameter compared with `self`

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, please change the locations of examples as per the following:

`ivy/functional/ivy/general.py`:
 1. `With Number input:` We can replace the int and float examples with Number. Also, we can remove the list and tuple examples. Apologies about this change but the current version of the guide suggests it this way.
 2.  `With ivy.Array input:`
 3. `With ivy.NativeArray input:`
 4.  `With a mix of ivy.Array and ivy.NativeArrayinstances:`
 5. `With one ivy.Container input:`
 6. `With multiple ivy.Container input:`
 7. `With a mix of ivy.Array and ivy.Container instances:`
 
`ivy/array/general.py`:
1.`With ivy.Array instance method:`
2.`With a mix of ivy.Array and ivy.NativeArray instance method:`

`ivy/container/general.py`:
1. `With one ivy.Container instances:`
2. `With multiple ivy.Container instances:`
1 and 2 are to be added inside the instance method.
3. `With one ivy.Container input:`
4. `With multiple ivy.Container input:`
3 and 4 are to be added to the static method.
We'll look into this after the previous ones are done.

Please remove any examples which are not mentioned in the list for a particular file. 
Once you're done with these changes, we'll look into the `ivy.Container` functions. Thanks 🙂

VedPatwardhan commented on ivy/container/general.py:
```diff
@@ -50,3 +51,106 @@ def clip_vector_norm(
+        x,
+        equality_matrix: bool = False
+    ):
+        """Determines whether the inputs are all equal.
```
The docstring before the examples need not have the description of the parameters and what's returned, you can refer [this](https://github.com/unifyai/ivy/blob/master/ivy/container/elementwise.py#L139) docstring for the same. Similarly, you can refer [this](https://github.com/unifyai/ivy/blob/master/ivy/container/elementwise.py#L82) docstring to see how to add the docstring for the static function you'll be adding.

VedPatwardhan commented on ivy/array/general.py:
```diff
@@ -1,8 +1,64 @@
+            x,
+            equality_matrix: bool = False
+    ):
+        """Determines whether the inputs are all equal.
```
The docstring before the examples need not have the description of the parameters and what's returned, you can refer [this](https://github.com/unifyai/ivy/blob/master/ivy/array/elementwise.py#L34) docstring for the same.

VedPatwardhan commented on ivy/functional/ivy/general.py:
```diff
@@ -524,6 +524,136 @@ def all_equal(
+        b: false
+    }
+
+    Instance Method Examples
```
We don't need the instance method examples in this file, please remove them

VedPatwardhan commented on ivy/container/general.py:
```diff
@@ -50,3 +51,106 @@ def clip_vector_norm(
+            Boolean, whether or not the inputs are equal, or matrix array of booleans
+            if equality_matrix=True is set.
+
+        Instance Method Examples
```
Both examples seem to be doing the same thing, please remove one example from each so that we have one example for `With one ivy.Container instances` and one example for `With multiple ivy.Container instances`

VedPatwardhan replied:
Also, the header can simply be `Examples`

VedPatwardhan commented on ivy/array/general.py:
```diff
@@ -1,8 +1,64 @@
+            Boolean, whether or not the inputs are equal, or matrix array of booleans
+            if equality_matrix=True is set.
+
+        Instance Method Examples
```
Both examples seem to be doing the same thing, please remove one example from each so that we have one example for `With one ivy.Array instance method` and one example for `With a mix of ivy.Array and ivy.NativeArray instance method`

VedPatwardhan replied:
Also, the header can simply be `Examples`

VedPatwardhan commented on ivy/functional/ivy/general.py:
```diff
@@ -524,6 +524,136 @@ def all_equal(
         Boolean, whether or not the inputs are equal, or matrix array of booleans if
         equality_matrix=True is set.
 
+    Functional Examples
```
The header can simply be `Examples`

VedPatwardhan replied:
Please also update the examples so that we have one example for each of `With Number inputs`, `With ivy.Array inputs`, `With ivy.NativeArray inputs` and so on. Also please make sure that one of the examples uses `equality_matrix=True`

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102, great work!
1. The next step is to add the container methods, two methods need to be added for the same, a static method and an instance method, as the instance method is already added, you'll have to add a static method to the `ivy/container/general.py` file, you can refer [this](https://github.com/unifyai/ivy/blob/master/ivy/container/elementwise.py#L71) function for the same. 
2. Next, you need to update the `all_equal` instance method to use this static method inside it, you can refer [this](https://github.com/unifyai/ivy/blob/master/ivy/container/elementwise.py#L128) function to see how that's done.
3. Then you'll have to add examples for using the static method too, in the form `ivy.Container.static_all_equal`.

There are some other changes to be made, I have pointed them out near the lines of code concerned, please have a look

davidlee1102 commented:
Hi @VedPatwardhan,
Thank you for your advice, I have been pushed the new fixed code; however, there were struggles that I want to ask you 
1. I have added but dont know that the docstring is correct or not, so please take a look on that and give me an advice
2.  I have added, let check again
3.  About `ivy.Container.static_all_equal`, it doesnt have, so I have tried `ivy.Container.equal`, let take a look again
4. The rest of problem which marked as resolved has been solved
Thank you
David

VedPatwardhan reviewed (changes requested):
Hi @davidlee1102,
1. The docstrings look good to me, we'll see if they fail the tests at the time of merging.
2. Great.
3. About the static method, according to the task guide, if the function doesn't exist, then that needs to be added, so feel free to add the static function with the name specified in the previous review referring to the function mentioned in the same review.
4. Great.

Let me know if you have any questions. Thanks 🙂

davidlee1102 commented:
Hi @VedPatwardhan ,
I have done all requested, let take a look again after merging this issue
1. Thank you for notifying me
2. Thank you 
3. I have added, so let take a look again
4. Thank you
David
